Created new Knative Services section (#1972)
knative/docs: docs/serving/using-knative-services.md

---
title: "Using Knative Services"
weight: 01
linkTitle: "Using Knative Services"
type: "docs"
---

Knative services are used to deploy an application. Knative Services attempt to model a more logical definition of a Service or Application. To achieve this, Services use the concept of Revisions which are each backed by a deployment with 2 Kubernetes Services.   

For more information about Kubernetes Services, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/services-networking/service/).   

Each Knative Service is defined by a Route and a Configuration, which have the same name as the Service, contained in a YAML file.

## Before you begin

To use Knative Services, you will need:

> * A Kubernetes cluster with [Knative installed](https://knative.dev/docs/install/index.html).

Reviewer: I think it would be beneficial to add a conditional statement here with a pointer to "configuring custom domains" so that its clear that there is additional work to just deploying a service before you can access it externally.

This YAML file specifies metadata about the application, points to the hosted image of the app and allows the Service to be configured.

This guide uses the [Hello World sample app in Go](https://knative.dev/docs/serving/samples/hello-world/helloworld-go) to demonstrate the structure of a Service YAML file and the basic workflow for deploying an app. These steps can be adapted for your own application if you have an image of it available on Docker Hub, Google Container Registry, or another container image registry.

The Hello World sample app does the following:
1. Reads an environment variable, `TARGET`, from the configuration .yaml file
1. Prints `Hello World: \${TARGET}!`. If `TARGET` is not defined, it will print `NOT SPECIFIED`.

### Procedure

1. Create a new file named `service.yaml` containing the following information.

```yml
apiVersion: serving.knative.dev/v1 # (1)
kind: Service
metadata:
 name: helloworld-go # (2)
 namespace: default # (3)
spec:
 template:
  spec:
   containers:
    - image: gcr.io/knative-samples/helloworld-go # (4)
     env:
      - name: TARGET # (5)
       value: "Go Sample v1"
```

> **(1)** Current version of Knative   

Reviewer: For long term maintenance, can we avoid qualifying the lines of code and just use a list of some sort? [Example here](https://knative.dev/development/eventing/accessing-traces/#configuration-options) (a definition list would work great too).
That way if the YAML changes (ie new or removed lines) then we can avoid updating numbers in the codes and the list. [I like to avoid and reserve numbered lists for only items that require some type of linear order (due to the limits of markdown, the numbers below end up looking like lists (rather than key value mappings)].

**(2)** The name of the app   
**(3)** The namespace the app will use   
**(4)** The URL to the image of the app   
**(5)** The environment variable printed out by the sample app

> Note: If you’re deploying an image of your own app, update the name of the app and the URL of the image accordingly.

## Creating a private cluster-local Service
By default services deployed through Knative are published to an external IP

> address, making them public services on a public IP address and with a

Reviewer: I think that if we use active voice here, we can address @dgerd's point too, for example:

```suggestion
address. After you configure your [custom domain](./using-a-custom-domain.md),  you can publicly access your services.
```

> [public URL](./using-a-custom-domain.md).

Reviewer: ```suggestion
```

While this is useful for services that need to be accessible from outside of the
cluster, frequently you may be building a backend service which should not be
available off-cluster.

Knative provides two ways to enable private services which are only available
inside the cluster:

> 1. To make all services only available only from within the cluster, change the default domain to

Reviewer: Similar to the comment above. Since these are two options and not a set of linear steps, 'bulleted lists' work better here.

   `svc.cluster.local` by
   [editing the `config-domain` config map](./using-a-custom-domain.md). This
   will change all services deployed through Knative to only be published to the
   cluster, none will be available off-cluster.

> 1. To make an individual service cluster-local, the service or route can be

Reviewer: "cluster-local" appears to be a visibility label that is Knative specific but Kubernetes refers to "both "cluster-private-IP" and "cluster-internal IP" terms". For consistency we should try to use the same terminology (versus saying "to be cluster-local")

For example: You can configure a Knative service to use a "cluster-private-IP"/"cluster-internal IP" if you add the `cluster-local` label to that service.

not on the public Internet), you can apply the
`serving.knative.dev/visibility=cluster-local` label to the Service object.

To label the Service:

```shell
kubectl label kservice ${KSVC_NAME} serving.knative.dev/visibility=cluster-local
```

To label a route:

```shell
kubectl label route ${ROUTE_NAME} serving.knative.dev/visibility=cluster-local
```

For example, you can deploy the [Hello World sample](./samples/helloworld-go)
and then convert it to be an cluster-local service by labeling the service:

```shell
kubectl label kservice helloworld-go serving.knative.dev/visibility=cluster-local
```

You can then verify that the change has been made by verifying the URL for the
helloworld-go service:

```shell
kubectl get ksvc helloworld-go

NAME            URL                                              LATESTCREATED         LATESTREADY           READY   REASON
helloworld-go   http://helloworld-go.default.svc.cluster.local   helloworld-go-2bz5l   helloworld-go-2bz5l   True
```

The service returns the a URL with the `svc.cluster.local` domain, indicating
the service is only available in the cluster local network.

> ## Deploying an application using Knative Services

Reviewer: From here down i think ALL but "Modifying Knative Services", is out of scope for this "Creating a Knative service" task and should move back into the "Getting Started" guide.

---
title: "User-created Knative Services"
weight: 01
linkTitle: "User-created Knative Services"
type: "docs"
---

> Knative services are Kubernetes Services that a user creates to deploy an application. Each Knative Service is defined by a Route and a Configuration, which have the same name as the Service, contained in a YAML file.

Reviewer 2: This first sentence is not quite right. Knative Services are backed by multiple Kubernetes "Services". This is because the Kubernetes Service are essentially load balancers over a Deployment. See https://kubernetes.io/docs/concepts/services-networking/service/ 

The Knative Service attempts to model a more logical definition of a Service or Application. To do so it has the concept of Revisions which are each backed by a deployment with 2 K8s Services.

## Before you begin

To use Knative Services, you will need:
* A Kubernetes cluster with [Knative installed](https://knative.dev/docs/install/index.html).

## Creating a Knative Service
To create an application, you need to create a YAML file that defines a Knative Service.

> This YAML file specifies metadata about the application, points to the hosted image of the app for deployment, and allows the deployment to be configured.

Reviewer 2: The word "deployment" here might be confusing as it is 
```suggestion
This YAML file specifies metadata about the application, points to the hosted image of the app, and allows the service to be configured.
```

The word deployment here might be confusing as it is overloaded with K8s deployments.

> This guide uses the [Hello World sample app in Go](https://knative.dev/docs/serving/samples/hello-world/helloworld-go) to demonstrate the structure of a Service YAML file and the basic workflow for deploying an app, but these steps can be adapted for your own application if you have an image of it available on Docker Hub, Google Container Registry, or another container image registry.

Reviewer 2: ```suggestion
This guide uses the [Hello World sample app in Go](https://knative.dev/docs/serving/samples/hello-world/helloworld-go) to demonstrate the structure of a Service YAML file and the basic workflow for deploying an app. These steps can be adapted for your own application if you have an image of it available on Docker Hub, Google Container Registry, or another container image registry.
```

The Hello World sample app does the following:

> 1. reads in an env variable, TARGET, from the configuration .yaml file

Reviewer 2: ```suggestion
1. reads in an environment variable, TARGET, from the configuration YAML file
```

> 2. prints “Hello World: \${TARGET}!“. If TARGET isn’t defined, it will print “NOT SPECIFIED”.

Reviewer 2: ```suggestion
1. prints `Hello World: \${TARGET}!`. If TARGET isn’t defined, it will print `NOT SPECIFIED`.
```

### Procedure

> 1. Create a new file named `service.yaml` containing the following information.

Reviewer 2: To make it easier to follow you could make this a [HereDoc](https://en.wikipedia.org/wiki/Here_document) that outputs to service.yaml so all I need to do is copy the command and press enter.

Author: First time I've heard of this :slightly_smiling_face: @RichieEscarez do you know if this is something we're using and if so, do we have any contributor docs for how to do this? Thanks.

Reviewer: @dgerd is this an example of what your are proposing (HereDoc), here in one of our other topics: https://knative.dev/docs/eventing/samples/kafka/channel/#creating-a-kafkachannel-channel-crd ?

ie. Where you wrap the [YAML] with with the following so that users can edit, apply, and deploy in a single command?

```
cat <<-EOF | kubectl apply -f -
---
[ALL_THE_YAML_GETS_COPIED_HERE?]
EOF
```

Question: This syntax does not create a copy of that file locally for the user, right? Is there a way to also create a local file?

@abrennan89 today, we have the [`readfile` shortcode](https://github.com/knative/website/blob/master/layouts/shortcodes/readfile.md) that will copy in an actual code file (at markdown build time) so that you dont have to manually copy and hardcode them into the markdown. Example here https://knative.dev/smoketest/

`{{< readfile file="./docs/serving/samples/hello-world/helloworld-go/service.yaml" code="true" lang="yaml" >}}`

Per @dgred's feedback, maybe I could open a PR to add an option to `readfile` (another flag: `heredoc="true"`) so that it automatically adds the "HereDoc" syntax around that code?

Reviewer 2: > This syntax does not create a copy of that file locally for the user, right? Is there a way to also create a local file?

This syntax does not. `tee` is a common command that can be inserted between to make a local copy. Alternatively the file creation through a heredoc and the application through kubectl could be broken into two steps if storing the file locally is desired.

```yml

> apiVersion: serving.knative.dev/v1 (1)

Reviewer 2: ```suggestion
apiVersion: serving.knative.dev/v1 # (1)
```

Make this and below comments so it copy/paste works.

kind: Service
metadata:
  name: helloworld-go (2)
  namespace: default (3)
spec:
  template:
    spec:
      containers:
        - image: gcr.io/knative-samples/helloworld-go (4)
          env:
            - name: TARGET (5)
              value: "Go Sample v1"
```
**(1)** Current version of Knative   
**(2)** The name of the app   
**(3)** The namespace the app will use   
**(4)** The URL to the image of the app   
**(5)** The environment variable printed out by the sample app

**NOTE:** If you’re deploying an image of your own app, update the name of the app and the URL of the image accordingly.

## Creating a private cluster-local Service
By default services deployed through Knative are published to an external IP
address, making them public services on a public IP address and with a

> [public URL](./using-a-custom-domain.md).

Reviewer 2: Probably worth noting here that the public URL out of the box is "example.com" which is non-functioning as it does not point to the public IP address.

Author: I think this is covered in / under the scope of https://knative.dev/docs/serving/using-a-custom-domain/ maybe?

While this is useful for services that need to be accessible from outside of the
cluster, frequently you may be building a backend service which should not be
available off-cluster.

Knative provides two ways to enable private services which are only available
inside the cluster:

> 1. To make all services only cluster-local, change the default domain to

Reviewer 2: ```suggestion
1. To make all services available only from within the cluster, change the default domain to
```

   `svc.cluster.local` by
   [editing the `config-domain` config map](./using-a-custom-domain.md). This
   will change all services deployed through Knative to only be published to the

>    cluster, none will be available off-cluster.

Reviewer 2: Might be too much information for this guide, but you can also add this domain as part of a selector statement to apply this to some of the services.

Author: @RichieEscarez wdyt with regards to document scope?

Reviewer: I think it can work here if we remove/revert combining all that "deploying an application" content in this file.

@dgerd Question: is creating "cluster-local services" not a common task? the idea here is that we teach users how to create a Knative service with this task and the provide the option for configuring that task to be "private"? If its not going to be a typical task, we could instead use "progressive disclosure" and only provide a "statement and link" to point users to a separate sub-topic (ie revert the "cluster-local-route" topic).

1. To make an individual service cluster-local, the service or route can be
   labeled in such a way to prevent it from getting published to the external
   gateway.

### Label a service to be cluster-local

To configure a Service to only be available on the cluster-local network (and
not on the public Internet), you can apply the

> `serving.knative.dev/visibility=cluster-local` label to the Service or Route

Reviewer 2: Since this is "using knative services" you will want to apply it to the Service. The Service will propagate the label to the Route. Any label added to a Route that is owned by a Service will be overwritten.

```

The service returns the a URL with the `svc.cluster.local` domain, indicating
the service is only available in the cluster local network.

## Deploying an application using Knative Services

To deploy an application, you must apply the `service.yaml` file.

### Procedure
1. Navigate to the directory where the `service.yaml` file is contained.
2. Deploy the application by applying the `service.yaml` file.   
  ```
  kubectl apply --filename service.yaml
  ```

Now that service has been created and the application has been deployed, Knative will create a new immutable revision for this version of the application.   

Knative will also perform network programming to create a route, ingress, service, and load balancer for your application, and will automatically scale your pods up and down based on traffic, including inactive pods.

## Interacting with your app

To see if your application has been deployed successfully, you need the host URL and IP
address created by Knative.

**NOTE:** If your cluster is new, it can take some time before the Service is
assigned an external IP address.

1. To find the IP address for your Service, enter:

   ```shell

>    # In Knative 0.2.x and prior versions, the `knative-ingressgateway` service was used instead of `istio-ingressgateway`.

Reviewer 2: 0.2.x is so old now that I don't think it is worth the confusion here.

   INGRESSGATEWAY=knative-ingressgateway

   # The use of `knative-ingressgateway` is deprecated in Knative v0.3.x.
   # Use `istio-ingressgateway` instead, since `knative-ingressgateway`

>    # will be removed in Knative v0.4.

Reviewer 2: Again these versions are old enough now that I would remove this comment..

```shell
   export IP_ADDRESS=$(kubectl get node  --output 'jsonpath={.items[0].status.addresses[0].address}'):$(kubectl get svc $INGRESSGATEWAY --namespace istio-system   --output 'jsonpath={.spec.ports[?(@.port==80)].nodePort}')
```

1. To find the host URL for your service, enter:

   ```shell
   kubectl get route helloworld-go  --output=custom-columns=NAME:.metadata.name,URL:.status.url
   ```

   The command will return the following:

   ```shell
   NAME                URL
   helloworld-go       http://helloworld-go.default.example.com
   ```

   > **NOTE:** By default, Knative uses the `example.com` domain. To configure a
   > custom DNS domain, see
   > [Using a Custom Domain](../serving/using-a-custom-domain.md).

   If you changed the name from `helloworld-go` to something else when creating
   the `.yaml` file, replace `helloworld-go` in the above commands with the name
   you entered.

1. Now you can make a request to your app and see the results. Replace
   `IP_ADDRESS` with the `EXTERNAL-IP` you wrote down, and replace
   `helloworld-go.default.example.com` with the domain returned in the previous
   step.

   ```shell

>    curl -H "Host: helloworld-go.default.example.com" http://${IP_ADDRESS}

Reviewer 2: It is probably worth explaining why we have to inject the host header here and that this is not necessary if you configure a domain associated with your IP.
